src/App.js

```diff
@@ -27,42 +27,80 @@ const generateSquares = () => {
 
 const App = () => {
 
-  // This starts state off as a 2D array of JS objects with
-  // empty value and unique ids.
-  const [squares, setSquares] = useState(generateSquares());
+  const [squares, setSquares] = useState(generateSquares()); // squares is created here
+  const [player, setPlayer] = useState(PLAYER_1);
+  const [winner, setWinner] = useState(`...`);
 
-  // Wave 2
-  // You will need to create a method to change the square 
-  //   When it is clicked on.
-  //   Then pass it into the squares as a callback
+  // until the new render has occurred, the squares reported in onClickCallback are the same squares set up by the previous render
+  const onClickCallback = (id) => {
+    const updatedSquares = generateSquares();
 
+    if (winner !== '...') return; // Don't continue updating after winner is declared
+    
+    for (let row = 0; row < 3; row += 1) {
+      for (let col = 0; col < 3; col += 1) {
+        if (updatedSquares[row][col].id === id && squares[row][col].value === '') {
+        updatedSquares[row][col].value = player;
+        setPlayer(player === PLAYER_1 ? PLAYER_2 : PLAYER_1);
+        }
+        else {
+          updatedSquares[row][col].value = squares[row][col].value
+        }
+      }
+    }
+
+    setSquares(updatedSquares);
+    checkForWinner(updatedSquares);
+  }
 
-  const checkForWinner = () => {
-    // Complete in Wave 3
-    // You will need to:
-    // 1. Go accross each row to see if 
-    //    3 squares in the same row match
-    //    i.e. same value
-    // 2. Go down each column to see if
-    //    3 squares in each column match
-    // 3. Go across each diagonal to see if 
-    //    all three squares have the same value.
+  const checkForWinner = (squares) => {
```

Comment:
Using a loop here might have simplified logic a bit but having a bit `if`/`else` definitely works well as a way to take care of this since you know you have a constant number of squares.  :smiley: